### ISSUE TITLE
Rule idea: declaration-nested-properties

### ISSUE THREAD

**Issue by s10wen** [kristerkari/stylelint-scss#44]
From https://github.com/stylelint/stylelint/issues/1386#issuecomment-223788453 @jeddy3 suggested I post this here.

**Comment by kristerkari** [comment]
Good idea 👍

**Comment by dryoma** [comment]
Getting down to this at last.

And the question. Do we really want the "always" option here? That would mean this rule will have to enforce nesting properties that... which ones exactly?

1) Definitely not those with a `-` in a name, 'cause `text-align` and stuff. Or no?

```scss
text: {
  align: center;
  decoration: underline;
  indent: 2em;
}
```

2) Or should we probably create a list of rules that are not allowed in non-nested form? E.g. `[ /^margin-/, /^flex-/, /^background-/, etc.]`, and if the parser meets one of them it yields something like `Expected "margin-left" to be in a form of a nested rule`.

3) Or just `"declaration-no-nested-properties`?

**Comment by s10wen** [comment]
@dryoma interesting, regarding `2)` are you thinking of setting defaults, or have the user set them? Also wondering what would be the best approach, blacklist / whitelist, or give the user the option to set them, then in the docs show a suggestion.

**Comment by kristerkari** [comment]
3) would be the easiest to implement, but yeah 2) could also work as set of whitelisted properties

**Comment by dryoma** [comment]
If we go with whitelist/blacklist, than no defaults are possible by the nature of stylelint. I was hoping that someone could point at the list of properties that are usually/preferably use nesting in Sass, as Sass itself (according to [Sassmeister](http://www.sassmeister.com/)) does'n seem to give two moos about what is nested:

```scss
a {
  may: {
    the: {
      force: be with you;
    }
  }
}
```
```css
a {
  may-the-force: be with you;
}
```

**Comment by kristerkari** [comment]
I guess if Sass does not care about what you nest, maybe we should not either -- and just skip option 2.

So we could do 3), or 1) that would have `"always" ` option to enforce nesting for any property that has `-` in the name.

**Comment by dryoma** [comment]
If we go with 1), then is this:

```scss
a {
  display: block;
  margin: {
    left: 10px;
  }
}
```
acceptable? I.e. nesting just one prop. Maybe this rule must be triggered only if two or more props with the same prefix (`margin-`) are met?

**Comment by kristerkari** [comment]
or maybe require everything to be nested by default and add  an option `except: first`?

**Comment by dryoma** [comment]
So, with `"always"`: 
```scss
a {
  margin-left: 10px; // a warning
  animation-name: weeee; // a warning
  animation-delay: 0.1s; // a warning
  text: {
    align: justify; // ok
  }
} 
```
with `[ "always", {except: "only-of-namespace"} ]` 

```scss
a {
  margin-left: 10px; // get off, I'm the only child!
  animation-name: weeee; // a warning
  animation-delay: 0.1s; // a warning
  text: {
    align: justify; // warning, as "except" options reverse corresponding main options
  }
} 
```
Like so?

Also, how should we name that `except` option? It is "the only property with certain prefix/of certain namespace". `"only-of-namespace"`?

p.s.
```scss
.funky {
  font: 20px/24px fantasy {
    weight: bold;
  }
}
```
makes me wanna go nuts rignt now :'(

**Comment by dryoma** [comment]
@s10wen the rules are available in 1.4.1 (at last)

**Comment by s10wen** [comment]
@dryoma niiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiice and <3 the example at https://github.com/kristerkari/stylelint-scss/issues/44#issuecomment-235237158 the force must be strong with you 👍
